File: 78-subsets/subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []
        def dfs(i, comb):
            if i == len(nums):
                res.append(list(comb))
                return
            dfs(i+1, comb)
            comb.append(nums[i])
            dfs(i+1, comb)
            comb.pop(-1)
        dfs(0, [])
        return res

File: 78-subsets/subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        dfs(0, new ArrayList<>(), res, nums);
        return res;
    }

    private void dfs(int i, List<Integer> comb, List<List<Integer>>res, int[] nums){
        if(i == nums.length){
            res.add(new ArrayList(comb));
            return;
        }

        comb.add(nums[i]);
        dfs(i+1, comb, res, nums);
        comb.remove(comb.size() - 1);
        dfs(i+1, comb, res, nums);
    }
}